src/language-switcher/widget/widget.php

```diff
@@ -0,0 +1,337 @@
+<?php
+namespace WPML\PB\Elementor\LanguageSwitcher;
+
+use Elementor\Controls_Manager;
+use Elementor\Group_Control_Typography;
+use Elementor\Scheme_Color;
+use Elementor\Widget_Base;
+
+class Widget extends Widget_Base {
```

Comment:
@strategio If you really want to test the code properly then I think you need to remove all the business logic out of this class and use an adaptor pattern. The should make this `Widget` very simple so it doesn't have any logic that you need to test with unit tests. Then you can fully test the adaptor.

```php
__construct( Adaptor $adaptor ) {
   parent::__construct();
    $this->adaptor = $adaptor;
    $adaptor->target( $this );
}

public function get_name() {
    return $this->adaptor->getName();
}

...
...

protected function _register_controls() {
    $this->adaptor->registerControls();
}

```

Then in the Adaptor you call `$this->target->start_controls_section( .... )`, etc, which you can then test to confirm the target is called with the correct data.

Reply:
Thanks for the suggestion @brucepearson. This is something I was thinking but I was unsure if all the methods used in `_register_controls` and `render` were public (I am talking of `get_settings_for_display`, `add_render_attribute`, `add_control`, ...). It looks like everything could be accessed by the adaptor. I will try to implement this!